### ISSUE TITLE
Cells not running after an exception caught with ipython custom exception handler

### ISSUE THREAD

**Issue by yogevyuval** [nteract/papermill#518]
When configuring a custom exception handler in iPython with: `ip.set_custom_exc` inside a notebook there is a strange and buggy behaviour in papermill.

The notebook runs until the cell the raises an exception (that is handled by the custom handler), and then the notebook return with the cells after the current cell marked as "Pending". No exception is raised and the next cells are not executed.

Example:

Cell 1 (running ok):
```
x = 5
print(x)
```

Cell 2 (custom exception):
```
class MyException(BaseException):
    pass

ip = get_ipython()
def general_exception_handler(self, etype, exc, tb, tb_offset=None):
    print("special handle")

ip.set_custom_exc((MyException,), general_exception_handler)
raise MyException('ads')
```

Cell 3:
```
# This does not raise an exception because it is not running, cell status is "pending"
1 / 0
```

**Comment by MSeal** [comment]
Sorry I haven't been able to get to all the issues quickly here. I believe this is actually an ipython issue (or a non-issue) and not a papermill one, and I'll try to explain why here.

So what the kernel is sending is as an execute reply for Cell2 as:
```
{
  "header": {
    "msg_id": "ae227f96-5f3c95a661ffc11124ed4a82_10",
    "msg_type": "execute_reply",
    "username": "mseal",
    "session": "ae227f96-5f3c95a661ffc11124ed4a82",
    "date": "2020-07-01T18:53:28.372825Z",
    "version": "5.3"
  },
  "msg_id": "ae227f96-5f3c95a661ffc11124ed4a82_10",
  "msg_type": "execute_reply",
  "parent_header": {
    "msg_id": "7971df3a04ce4c3582e156da0ddb60ad",
    "username": "username",
    "session": "628f1ab3b1e5450d856c704580171064",
    "msg_type": "execute_request",
    "version": "5.2",
    "date": "2020-07-01T18:53:28.363505Z"
  },
  "metadata": {
    "started": "2020-07-01T18:53:28.363812Z",
    "dependencies_met": true,
    "engine": "163804de-1ea9-4435-bd99-5da89a1bbbf9",
    "status": "error"
  },
  "content": {
    "status": "error",
    "traceback": [],
    "ename": "MyException",
    "evalue": "ads",
    "engine_info": {
      "engine_uuid": "163804de-1ea9-4435-bd99-5da89a1bbbf9",
      "engine_id": -1,
      "method": "execute"
    },
    "execution_count": 1,
    "user_expressions": {},
    "payload": []
  },
  "buffers": [],
  "channel": "shell"
}
```
which is indicating that the cell as failed with a status `error` and papermill then stops execution.

From the [set_custom_exc docs](https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.interactiveshell.html) it doesn't indicate if the exception handler will or will not reraise, though the quote about custom raised exceptions will be ignored seems to indicate that it will always reraise the original exception after the handler is fired. I believe this means that the behavior as seen is expected.

**Comment by yogevyuval** [comment]
@MSeal Thanks for the response!

I see what you way, but that raises another questions:
1. Will I see this error status on the cell in papermill result?
2. Does this mean I have to check each papermill result to see if all cells executed succeffully? I was under the impression that a return value (no exceptions) means success

**Comment by MSeal** [comment]
> Will I see this error status on the cell in papermill result?

You're supposed to.. I think because no exception is raised but an error status update is still being sent papermill is incorrectly classifying it as a successful stop instead of a failed execution.

 > Does this mean I have to check each papermill result to see if all cells executed succeffully? I was under the impression that a return value (no exceptions) means success

When failing a notebook it should raise an exception or exit the CLI with a non-zero status code.

I'll need to look deeper into nbclient and papermill to determine exactly what would need to be changed to generate an exception here.

**Comment by yogevyuval** [comment]
@MSeal 
After checking a bit more I do believe that there is some unwanted behaviour here.
The cell that is raising the exception is returning as:

```
{
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "tags": [],
        "papermill": {
          "exception": True,
          "start_time": "2020-07-05T10:25:55.016905",
          "end_time": "2020-07-05T10:25:55.025990",
          "duration": 0.009085,
          "status": "failed"
        },
        "execution": {
          "iopub.status.busy": "2020-07-05T10:25:55.022687Z",
          "iopub.execute_input": "2020-07-05T10:25:55.023432Z",
          "shell.execute_reply": "2020-07-05T10:25:55.025135Z",
          "iopub.status.idle": "2020-07-05T10:25:55.025820Z"
        }
      }
```

Note that status is **failed** and 'exception' is **true**.

The problem here is that papermill returns with no exceptions, but rather it marks it on this specific cell metadata. This means that anyone that is running papermill has to check every cell output and metadata to be sure the notebook has actually run successfully, instead of relying on a general try-catch.

**Comment by MSeal** [comment]
No, as I stated it's a bug in how nbclient (which is doing the lifting for papermill here) is handling the logic for this situation. If you don't include the ipython exception handler you do get a proper exception raised and a non-zero exit code. The code is falling into a normally legitimate condition path where it allows an exception but continues execution because of how the kernel is responding to the handler.

**Comment by MSeal** [comment]
It might take me a little bit to track down the fix, as I've got a lot of open issues and not enough maintainers looking at the issues.

**Comment by hekmati** [comment]
Hi @MSeal, was there any progress regarding this issue?

**Comment by enryH** [comment]
Another example for the above: An error is not raised when one uses the `%%time` cell magic from IPython:

```
%%time
raise ValueError("Something is wrong, stop.")
a = 1 # executed
```

and if I then want to use `a` later I would get a `NameError`:

```papermill.exceptions.PapermillExecutionError:
---------------------------------------------------------------------------
Exception encountered at "In [2]":
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-14-3f786850e387> in <module>
----> 1 a

NameError: name 'a' is not defined
```

without the  `%%time` execution would stop on encountering the `ValueError`
